test/sketch_test.cpp

```diff
@@ -65,57 +65,125 @@ TEST(SketchTestSuite, GIVENonlyIndexZeroUpdatedTHENitWorks) {
+        //Undetected sample error
+        sample_incorrect_failures++;
+      }
+    } catch (AllBucketsZeroException& e) {
+      //All buckets being 0 implies that the whole vector should be 0
+      bool vec_zero = true;
+      for (unsigned long j = 0; vec_zero && j < vec_size; j++) {
+        if (test_vec.get_entry(j) != 0) {
+          vec_zero = false;
+        }
+      }
+      if (!vec_zero) {
+        sample_incorrect_failures++;
+      }
+    } catch (NoGoodBucketException& e) {
+      //No good bucket
+      all_bucket_failures++;
+    } catch (MultipleQueryException& e) {
+      //Multiple queries shouldn't happen, but if we do get here fail test
+      FAIL() << e.what();
     }
   }
+  EXPECT_LE(sample_incorrect_failures, max_sample_fail_prob * num_sketches)
+    << "Sample incorrect " << sample_incorrect_failures << '/' << num_sketches
+    << " times (expected less than " << max_sample_fail_prob << ')';
+  EXPECT_LE(all_bucket_failures, max_bucket_fail_prob * num_sketches)
+    << "All buckets failed " << all_bucket_failures << '/' << num_sketches
+    << " times (expected less than " << max_bucket_fail_prob << ')';
 }
 
-TEST(SketchTestSuite, WHENmanySketchesAreSampledTHENallOfThemAreNonzero) {
-  srand(time(NULL));
-  int n = 100000;
-  int vec_size = 1000;
-  int num_updates = 100;
-  int times = 2;
+TEST(SketchTestSuite, TestSketchSample) {
+  srand (time(NULL));
+  test_sketch_sample(10000, 100, 100, 0.005, 0.005);
+  test_sketch_sample(1000, 1000, 1000, 0.001, 0.001);
+  test_sketch_sample(1000, 10000, 10000, 0.001, 0.001);
+}
 
-  while (times--) {
-    bool failed_flag = 0;
-    Testing_Vector test_vec = Testing_Vector(vec_size,num_updates);
-    for (int i=0;i<n;++i) {
-      Sketch sketch = Sketch(vec_size,rand());
-      for (int j=0;j<num_updates;++j) {
-        sketch.update(test_vec.get_update(j));
+/**
+ * Make sure sketch addition works
+ */
+void test_sketch_addition(unsigned long num_sketches,
+  unsigned long vec_size, unsigned long num_updates,
+  double max_sample_fail_prob, double max_bucket_fail_prob)
+{
+  unsigned long all_bucket_failures = 0;
+  unsigned long sample_incorrect_failures = 0;
+  for (int i = 0; i < num_sketches; i++){
+    const long seed = rand();
+    Sketch sketch1 = Sketch(vec_size, seed);
+    Sketch sketch2 = Sketch(vec_size, seed);
+    Testing_Vector test_vec1 = Testing_Vector(vec_size, num_updates);
+    Testing_Vector test_vec2 = Testing_Vector(vec_size, num_updates);
+
+    for (unsigned long j = 0; j < num_updates; j++){
+      sketch1.update(test_vec1.get_update(j));
+      sketch2.update(test_vec2.get_update(j));
+    }
+    Sketch sketchsum = sketch1 + sketch2;
+    try {
+      Update res = sketchsum.query();
+      ASSERT_NE(res.delta, 0) << "Sample is zero";
+      ASSERT_LT(res.index, vec_size) << "Sampled index out of bounds";
+      if (res.delta != test_vec1.get_entry(res.index) + test_vec2.get_entry(res.index)) {
+        sample_incorrect_failures++;
```

Comment:
we should fail immediately here. if the provided delta is not the sum of the vectors, then our addition is wrong.

Reply:
I'm pretty sure it's possible to get a false positive.